### PR TITLE
0.10.0-pre

### PR DESCRIPTION
**Pull request by gdotdesign**
This PR contains a lot of changes and features for the next release.

---
### Providers

Providers got overhauled in this release, it is possible to write a provider entirely in Mint.

Providers now work like stores they can have states, constants and getters, the `next` keyword can also be used.

All providers need to implement the `update` function which is called any time the subscriptions for that provider change. This allows changing the state accordingly, for example by adding or removing event listeners from entities (window, document, etc...)

If you are curious about them you can check the currently available providers here: https://github.com/mint-lang/mint/tree/master/core/source/Provider

### Directives

This release adds the `@svg`, `@format` and `@documentation` directives.

Directives are processed at compile time. They interact with entities both in the code or outside of them (for example files).

The `@svg` directive embeds the SVG at the given path (relative to the file) as `Html` so it can be used directly in components.

```
<div>
  <{ @svg(../path/to/file) }>
</div>
```

The `@format` directive returns the formatted string version of the code it's given and the result of the code in a tuple `Tuple(a, String)`. This is created to make it easy to display a simple example of Mint code:

```
try {
  {result, code} =
    @format {
      <div>
        "Hello There!"
      </div>
    }

  <div>
    <{ result }>

    <pre>
      <code>
        <{ code }>
      </code>
    </pre>
  </div>
}
```

The `@documentation` directive compiles the JSON documenation of the given entity into the resulting JavaScript code so it can be used to display information about the entity.

```
record Component {
  name : String
}

try {
  decoded =
    decode @documentation(Main) as Component

  decoded.name
}
```

### Enums

Enums now can have named parameters:

```
enum Item {
  Link(name : String, href : String)
}

Item::Link(name = "Home", href = "/")
```

This way there is no need to define a record for an enum item.

### Html

- Html expressions can now for attributes:

  ```
  <Field label=<{ "Hello" }>/>
  ```

- Html experssions can now used as experssions:

  ```
  try {
    lebel =
      <{ "Hello" }>

    <Field label={label}/>
  }
  ```

- Html expressions now supports multiple expressions (without needing Html expressions):

  ```
  <{
    "String"

    <div>
      "Hello"
    </div>

    if (Html.isNotEmpty(content)) {
      content
    } else {
      <{}>
    }
  }>
  ```

### Record Update

Record update now supports updating the result of an expression instead of just
updating a record pointing to a variable:

```
{ { a = "b" } | a = "x" } # { a = "x" }
```

### Constants

- Constants are now compiled into properties with static value instead of a
property getter.
- Constants now can have numbers in them: `const SOME_CONSTANT_01 = "A"`

### References

Component and DOM element references are property returning a `Maybe::Nothing`
if the reference cannot be found.

## Formatter

- Multiple operations can be auto formatted into multiple lines:

  ```
  "Hello" == "True" &&
    "a" == "b" &&
    "c" == "d"
  ```

- Case branches can now be forced to format as multiple by placing the expression in a new line:

  ```
  Maybe::Nothing =>
    "Hello There!"
  ```

- Constants can now be forced to format as multiple by placing the expression in a new line:

  ```
  const HELLO =
    "Hello There!"
  ```

- Html elements with only a string child will now be formatted in one line:

  ```
  <div>"Hello There!"</div>
  ```

- Html fragments and Html expressions are formatted more consistently

## CSS

Added the following properties:

- `overscroll-behavior`
- `overscroll-behavior-x`
- `overscroll-behavior-y`

### Added

- `Array.findByAndMap`
- `Dom.getElementsBySelector`
- `Dom.containsMaybe`
- `Dom.setAttribute`
- `Dom.getTagName`
- `Dom.getChildren`
- `Dom.getTextContent`
- `Dom.getActiveElement`
- `Dom.blurActiveElement`
- `Dom.getTextWidth`
- `Dom.getFocusableElements` 
- `Dom.focusFirst`
- `Dom.smoothScrollTo`
- `Dom.scrollTo`
- `Dom.getClientWidth`
- `Dom.getClientHeight`
- `Dom.getScrollLeft`
- `Dom.getScrollWidth`
- `Dom.getScrollTop`
- `Dom.getScrollHeight`
- `Dom.getTableOfContents`
- `Html.isNotEmpty`
- `Map.entries`
- `Promise.never1`
- `Promise.never2`
- `Promise.never3`
- `Promise.create`
- `String.isNotEmpty`
- `String.isBlank`
- `String.isNotBlank`
- `String.dropLeft`
- `String.parameterize`
- `Test.Context.`spyOn
- `Test.Context.`assertFunctionCalled
- `Test.Context.`assertFunctionNotCalled
- `Window.alert`
- `Window.isActiveURL`
- `Window.triggerHashJump`
- `Window.addEventListener`
- `Window.addMediaQueryListener`
- `Window.matchesMediaQuery`

### Changed

- `String.isAnagarm` -> `String.isAnagram`
- `Http.sendWithID` -> `Http.sendWithId`
- `Test.Html.assertCSSOf` -> `Test.Html.assertCssOf`
- `Dom.getAttribute` now returns `Maybe(String)`
- `Map` is now using arrays for the implementation #268
- `Set` is now using arrays for the implementation #268

### Fixed

- `Dom.getElementFromPoint` no longer results in a runtime error.

### Web Api Wrappers

- WebSocket - #285 #276
- AnimationFrame
- MutationOberver
- ResizeObserver
- IntersectionObserver

### New Providers

- Provider.ElementSize
- Provider.Intersection
- Provider.Keydown
- Provider.Keyup
- Provider.Mutation
- Provider.OutsideClick
- Provider.TabFocus
- Provider.Url
- Provider.WebSocket

### New Modules

- Validation

## Minor fixes

- Fix formatting of a global component with a comment
- Fix formatting of constant access
- Fix an issue of generating wrong CSS
- Fix formatting of number literals #263 #247
- Make sure `Tuple(Bool)` can be destructured in case branches
- CSS definitions can now have numbers in them
- The `format` command now formats using the given pattern
- The `Main` component cannot have properties and will result in a type error
- Updated scaffolding of a new project to include a head file

## Fixed issues

- #295 - Error message text does not fit in its container
- #252 - Not enough context in required props error
- #242 - Required property does not generate a type checker error
- #243 - Emit compilation error message to the CLI for `mint start`

### CODE REVIEW

File: core/source/Dom.mint [hunk 1]
```diff
@@ -286,4 +330,239 @@ module Dom {
...
+  Blurs the active element of the page.
+
+    Dom.blurActiveElement()
+  */
+  fun blurActiveElement : Promise(Never, Void) {
+    `document.activeElement && document.activeElement.blur()`
+  }
+
+  /*
+  Measures the given text with the given font using the canvas.
```
Comment (Sija) [inline core/source/Dom.mint:390]:
```suggestion
  Measures the given text width with the given font using the canvas.
```

File: core/source/Dom.mint [hunk 2]
```diff
@@ -286,4 +330,239 @@ module Dom {
...
+
+  /*
+  Scrolls the given element to the given position.
+
+    Dom.scrollTo(element, 10, 10)
+  */
+  fun scrollTo (element : Dom.Element, left : Number, top : Number) : Promise(Never, Void) {
+    `#{element}.scrollTo({
+      left: #{left},
+      top: #{top} })`
```
Comment (Sija) [inline core/source/Dom.mint:481]:
```suggestion
    `#{element}.scrollTo({
        left: #{left},
        top: #{top}
      })`
```

File: core/source/Promise.mint [hunk 3]
```diff
@@ -5,6 +5,30 @@ module Promise {
...
+  */
+  fun never1 (param1 : a) : Promise(Never, Void) {
+    Promise.resolve(void)
+  }
+
+  /*
+  Returns a resolved promise with `Void` which never fails with two
+  arguments which are ignored.
+  */
+  fun never2 (param1 : a, param2 : b ) : Promise(Never, Void) {
```
Comment (Sija) [inline core/source/Promise.mint:20]:
```suggestion
  fun never2 (param1 : a, param2 : b) : Promise(Never, Void) {
```

File: core/source/Promise.mint [hunk 4]
```diff
@@ -5,6 +5,30 @@ module Promise {
...
+    Promise.resolve(void)
+  }
+
+  /*
+  Returns a resolved promise with `Void` which never fails with three
+  arguments which are ignored.
+  */
+  fun never3 (param1 : a, param2 : b, param3 : c ) : Promise(Never, Void) {
```
Comment (Sija) [inline core/source/Promise.mint:28]:
```suggestion
  fun never3 (param1 : a, param2 : b, param3 : c) : Promise(Never, Void) {
```

File: core/source/String.mint [hunk 5]
```diff
@@ -213,4 +213,44 @@ module String {
...
   }
+
+  /*
+  Returns whether or not the string is not empty (only contains whitespace).
+
+    String.isNotEmpty("a") == true
+    String.isNotEmpty("   ") == false
+  */
+  fun isNotEmpty (string : String) : Bool {
+    !String.isEmpty(String.trim(string))
```
Comment (Sija) [inline core/source/String.mint:224]: IMO that should be 1:1 reversible with `String.isEmpty`:

```suggestion
    !String.isEmpty(string)
```

Desired behaviour should be a part of `String.isBlank` instead.

File: core/source/String.mint [hunk 6]
```diff
@@ -213,4 +213,44 @@ module String {
...
+    String.parameterize("HELLO THERE!!!") == "hello-there"
+    String.parameterize("-_!ASD!_-") == "asd"
+  */
+  fun parameterize (string : String) : String {
+    `
+    #{string}
+    .replace(/[^a-z0-9\-_]+/ig, '-')
+    .replace(/-{2,}/g, '-')
+    .replace(/^-|-$/i, '')
+    .toLowerCase()
```
Comment (Sija) [inline core/source/String.mint:253]:
```suggestion
      .replace(/[^a-z0-9\-_]+/ig, '-')
      .replace(/-{2,}/g, '-')
      .replace(/^-|-$/i, '')
      .toLowerCase()
```

File: spec/compilers_spec.cr [hunk 7]
```diff
@@ -1,21 +1,25 @@
 require "./spec_helper"
 
-Dir.glob("./spec/compilers/**/*").sort.each do |file|
-  it file do
-    # Read and separate sample from expected
-    sample, expected = File.read(file).split("-" * 80)
+Dir
+  .glob("./spec/compilers/**/*")
+  .select { |file| File.file?(file) }
+  .sort
```
Comment (Sija) [inline spec/compilers_spec.cr:6]:
```suggestion
  .select! { |file| File.file?(file) }
  .sort!
```

File: spec/formatting_spec.cr [hunk 8]
```diff
@@ -1,32 +1,36 @@
 require "./spec_helper"
 
-Dir.glob("./spec/formatters/**/*").each do |file|
-  it file do
-    # Read and separate sample from expected
-    sample, expected = File.read(file).split("-" * 80)
+Dir
+  .glob("./spec/formatters/**/*")
+  .select { |file| File.file?(file) }
+  .sort
```
Comment (Sija) [inline spec/formatting_spec.cr:6]:
```suggestion
  .select! { |file| File.file?(file) }
  .sort!
```

File: spec/type_checking_spec.cr [hunk 9]
```diff
@@ -1,55 +1,59 @@
...
-
-  contents.scan(/^\-+(\w+)?/m) do |match|
-    samples << {contents[position, match.begin.not_nil! - position], error}
-    position = match.end.not_nil!
-    error = match[1]?
-  end
+Dir
+  .glob("./spec/type_checking/**/*")
+  .select { |file| File.file?(file) }
+  .sort
```
Comment (Sija) [inline spec/type_checking_spec.cr:6]:
```suggestion
  .select! { |file| File.file?(file) }
  .sort!
```

File: src/ast/module_access.cr [hunk 10]
```diff
@@ -1,13 +1,14 @@
 module Mint
   class Ast
     class ModuleAccess < Node
-      getter variable, name
+      getter variable, name, constant
```
Comment (Sija) [inline src/ast/module_access.cr:4]:
```suggestion
      getter variable, name
      getter? constant
```

File: src/formatters/module_access.cr [hunk 11]
```diff
@@ -4,7 +4,14 @@ module Mint
       variable =
         format node.variable
 
-      "#{node.name}.#{variable}"
+      separator =
+        if node.constant
```
Comment (Sija) [inline src/formatters/module_access.cr:8]:
```suggestion
        if node.constant?
```

File: src/type_checkers/access.cr [hunk 12]
```diff
@@ -30,10 +30,29 @@ module Mint
...
+            when Ast::HtmlComponent
+              component_records
+                .find(&.first.name.==(ref.component.value))
+                .try do |entity|
+                  memo[variable.value] = entity.first
+                end
+            when Ast::HtmlElement
+              memo[variable.value] = ref
+            else
+              # ignore
```
Comment (Sija) [inline src/type_checkers/access.cr:45]:
```suggestion
```

File: src/type_checkers/component.cr [hunk 13]
```diff
@@ -32,6 +32,17 @@ module Mint
         fields[item.name.value] = static_type_signature(item)
       end
 
+      node.refs.each do |variable, ref|
+        case (ref)
```
Comment (Sija) [inline src/type_checkers/component.cr:36]:
```suggestion
        case ref
```

File: src/type_checkers/enum_destructuring.cr [hunk 14]
```diff
@@ -25,14 +25,36 @@ module Mint
...
+            .as(Ast::EnumRecordDefinition)
+            .fields
+
+        node.parameters.each do |item|
+          found = fields.find do |field|
+            case item
+            when Ast::TypeVariable
+              item.value == field.key.value
+            else
+              nil
```
Comment (Sija) [inline src/type_checkers/enum_destructuring.cr:42]:
```suggestion
```

File: core/source/Array.mint [hunk 15]
```diff
@@ -705,4 +705,31 @@ module Array {
...
+          return #{Maybe::Just(`value`)}
+        }
+      }
+
+      return #{Maybe::Nothing}
+    })()
+    `
+  }
+
+  fun reverseIf (condition : Bool, array : Array(a)) : Array(a) {
```
Comment (Sija) [inline core/source/Array.mint:728]: I'd keep the main receiver always as a first argument to allow chaining:

```elixir
[1, 2, 3]
|> reverseIf "foo" == "bar"
```

```suggestion
  fun reverseIf (array : Array(a), condition : Bool) : Array(a) {
```

ditto all other methods in stdlib (could be worth separate PR changing just that).

File: core/source/Dom.mint [hunk 16]
```diff
@@ -240,17 +268,33 @@ module Dom {
...
+  returns the element.
+
+    "a"
+    |> Dom.createElement
+    |> Dom.setAttribute("name", "test")
+  */
+  fun setAttribute (
+    attribute : String,
+    value : String,
+    element : Dom.Element
```
Comment (Sija) [inline core/source/Dom.mint:293]: ditto

```suggestion
    element : Dom.Element,
    attribute : String,
    value : String
```

File: core/source/Dom.mint [hunk 17]
```diff
@@ -286,4 +330,239 @@ module Dom {
...
+  /*
+  Smooth scroll the given element to the given position.
+
+    Dom.smoothScrollTo(element, 10, 10)
+  */
+  fun smoothScrollTo (element : Dom.Element, left : Number, top : Number) : Promise(Never, Void) {
+    `#{element}.scrollTo({
+      behavior: 'smooth',
+      left: #{left},
+      top: #{top} })`
```
Comment (Sija) [inline core/source/Dom.mint:470]:
```suggestion
    `#{element}.scrollTo({
        behavior: 'smooth',
        left: #{left},
        top: #{top}
      })`
```

File: core/source/Dom.mint [hunk 18]
```diff
@@ -286,4 +330,239 @@ module Dom {
...
+  /*
+  Returns the table of contents of the given element for the given selectors.
+
+    Dom.getTableOfContents("h1, h2, h3, h4", element) == [
+      {"h1", "The title of the page", "the-title-of-the-page"},
+      {"h2", "A subtitle of the page", "a-subtitle-of-the-page"},
+      {"h3", "A sub-subtitle of the page", "a-sub-subtitle-of-the-page"}
+    ]
+  */
+  fun getTableOfContents (selector : String, element : Dom.Element) : Array(Tuple(String, String, String)) {
```
Comment (Sija) [inline core/source/Dom.mint:547]: ditto

```suggestion
  fun getTableOfContents (element : Dom.Element, selector : String) : Array(Tuple(String, String, String)) {
```

File: core/source/Array.mint (hunk 15 above)
Comment (gdotdesign) [reply core/source/Array.mint:728]: The pipe operator `|>` puts the left side as the **last parameter**. 

Maybe you are confusing it with an other language? :D

Comment (Sija) [reply core/source/Array.mint:728]: Oh, I might :O IIRC Elixir passes it as a 1st argument so that's where I could the inspiration from :D

File: core/source/Dom.mint [hunk 19]
```diff
@@ -286,4 +330,241 @@ module Dom {
...
+  fun blurActiveElement : Promise(Never, Void) {
+    `document.activeElement && document.activeElement.blur()`
+  }
+
+  /*
+  Measures the given text width with the given font using the canvas.
+
+    Dom.measureText("20px sans-serif", "Hello There") = 300
+  */
+  fun measureText (font : String, text : String) : Number {
```
Comment (Sija) [inline core/source/Dom.mint:394]:
```suggestion
    Dom.getTextWidth("20px sans-serif", "Hello There") = 300
  */
  fun getTextWidth (font : String, text : String) : Number {
```

File: core/source/String.mint (hunk 5 above)
Comment (Sija) [reply core/source/String.mint:224]: @gdotdesign Still unresolved.

File: core/source/Validation.mint [hunk 20]
```diff
@@ -0,0 +1,201 @@
...
+  Returns the given error if the given string does not have at least the given
+  number of characters.
+
+    Validation.hasNumberOfCharacters(
+      "",
+      5,
+      {"zip", "Zip code does is not 5 characters!"}) ==
+        Maybe::Just({"zip", "Zip code does is not 5 characters!"})
+  */
+  fun hasNumberOfCharacters (
```
Comment (Sija) [inline core/source/Validation.mint:129]:
```suggestion
    Validation.hasMinimumNumberOfCharacters(
      "",
      5,
      {"zip", "Zip code does is not 5 characters or more!"}) ==
        Maybe::Just({"zip", "Zip code does is not 5 characters or more!"})
  */
  fun hasMinimumNumberOfCharacters (
```

File: src/compilers/component.cr [hunk 21]
```diff
@@ -174,13 +180,15 @@ module Mint
...
-          # If the user defined the same function the code goes after it.
-          if function && value
-            function.keep_name = true
+          function.keep_name = true if function
 
+          # If the user defined the same function the code goes after it.
+          if function && value.any?
             compile function, js.statements(value)
-          elsif !value.empty?
+          elsif value.any?
```
Comment (Sija) [inline src/compilers/component.cr:188]: `empty?` is in general more performant than `any?`.

```suggestion
          elsif !value.empty?
```

File: src/compilers/directives/svg.cr [hunk 22]
```diff
@@ -0,0 +1,46 @@
...
+    def parse_svg(contents)
+      document =
+        XML.parse(contents)
+
+      svg =
+        document.first_element_child
+
+      if svg
+        data =
+          svg.children.map(&.to_s).join("").strip
```
Comment (Sija) [inline src/compilers/directives/svg.cr:40]:
```suggestion
          svg.children.join.strip
```

File: src/compilers/variable.cr [hunk 23]
```diff
@@ -47,7 +47,18 @@ module Mint
...
+          when Ast::Component
+            ref =
+              parent
+                .refs
+                .find { |(ref, _)| ref.value == node.value }
+                .try { |(ref, _)| js.variable_of(ref) }
+
+            "this.#{ref}"
+          else
+            "SHOULD NOT HAPPEN"
```
Comment (Sija) [inline src/compilers/variable.cr:60]:
```suggestion
            raise "SHOULD NOT HAPPEN"
```

File: src/parsers/directives/format.cr [hunk 24]
```diff
@@ -0,0 +1,27 @@
+module Mint
+  class Parser
+    syntax_error FormatDirectiveExpectedOpeningBracket
+    syntax_error FormatDirectiveExpectedClosingBracket
+    syntax_error FormatDirectiveExpectedExpression
+
+    def format_directive : Ast::Directives::Format | Nil
```
Comment (Sija) [inline src/parsers/directives/format.cr:7]:
```suggestion
    def format_directive : Ast::Directives::Format?
```

File: src/parsers/directives/documentation.cr [hunk 25]
```diff
@@ -0,0 +1,25 @@
+module Mint
+  class Parser
+    syntax_error DocumentationDirectiveExpectedClosingParentheses
+    syntax_error DocumentationDirectiveExpectedOpeningParentheses
+    syntax_error DocumentationDirectiveExpectedEntity
+
+    def documentation_directive : Ast::Directives::Documentation | Nil
```
Comment (Sija) [inline src/parsers/directives/documentation.cr:7]:
```suggestion
    def documentation_directive : Ast::Directives::Documentation?
```

File: src/parsers/directives/svg.cr [hunk 26]
```diff
@@ -0,0 +1,28 @@
+module Mint
+  class Parser
+    syntax_error SvgDirectiveExpectedOpeningParentheses
+    syntax_error SvgDirectiveExpectedClosingParentheses
+    syntax_error SvgDirectiveExpectedPath
+
+    def svg_directive : Ast::Directives::Svg | Nil
```
Comment (Sija) [inline src/parsers/directives/svg.cr:7]:
```suggestion
    def svg_directive : Ast::Directives::Svg?
```

File: src/style_builder.cr [hunk 27]
```diff
@@ -117,6 +117,10 @@ module Mint
   class StyleBuilder
     class Selector < Hash(String, PropertyValue)
       getter id : String = Random::Secure.hex
+
+      def ==(other)
+        other.id == id
+      end
```
Comment (Sija) [inline src/style_builder.cr:123]:
```suggestion
      def_equals @id
```

File: src/compilers/component.cr (hunk 21 above)
Comment (gdotdesign) [reply src/compilers/component.cr:188]: Didn't know that, probably this should be a rule in Ameba or it should be optimized to check for emptiness and return early.

For me `any?` is more readable.

File: src/compilers/constant.cr [hunk 28]
```diff
@@ -1,13 +1,12 @@
...
-      name =
-        js.variable_of(node)
-
-      value =
-        compile node.value
-
-      js.get(name, "return #{value};")
+    def compile_constants(nodes : Array(Ast::Constant)) : Hash(String, String)
+      nodes
+        .select { |node| checked.includes?(node) }
```
Comment (Sija) [inline src/compilers/constant.cr:5]:
```suggestion
        .select(&.in?(checked))
```

File: core/source/String.mint [hunk 29]
```diff
@@ -213,4 +213,44 @@ module String {
...
+
+  /*
+  Returns whether or not the string is not empty (only contains whitespace).
+
+    String.isNotEmpty("a") == true
+    String.isNotEmpty("   ") == false
+  */
+  fun isNotEmpty (string : String) : Bool {
+    !String.isEmpty(String.trim(string))
+  }
```
Comment (Sija) [inline core/source/String.mint:225]:
```suggestion
  /*
  Returns whether or not the string is not empty (has no value).

    String.isNotEmpty("a") == true
    String.isNotEmpty("") == false
  */
  fun isNotEmpty (string : String) : Bool {
    !String.isEmpty(string)
  }
```

File: core/source/Validation.mint [hunk 30]
```diff
@@ -0,0 +1,201 @@
...
+
+  /*
+  Returns the given error when the given string is blank
+  (contains only whitespace).
+
+    Validation.isNotBlank("", {"name", "Name is empty!"}) ==
+      Maybe::Just({"name", "Name is empty!"})
+  */
+  fun isNotBlank (value : String, error : Tuple(String, String)) : Maybe(Tuple(String, String)) {
+    if (String.isNotEmpty(value)) {
```
Comment (Sija) [inline core/source/Validation.mint:60]:
```suggestion
    if (String.isNotBlank(value)) {
```

File: core/tests/tests/String.mint [hunk 31]
```diff
@@ -137,3 +137,32 @@ suite "String.trim" {
...
+
+  test "it returns false for fully empty string" {
+    String.isNotEmpty("") == false
+  }
+
+  test "it returns false for empty string (whitespace)" {
+    String.isNotEmpty(
+      "
+      ") == false
+  }
```
Comment (Sija) [inline core/tests/tests/String.mint:154]:
```suggestion
  test "it returns true for empty string (whitespace)" {
    String.isNotEmpty("  ") == true
  }
```

Comment (gdotdesign) [reply core/tests/tests/String.mint]: I've added `String.isBlank` and `String.isNotBlank` :smile: :+1:

File: core/source/String.mint [hunk 32]
```diff
@@ -38,6 +38,37 @@ module String {
     string == ""
   }
 
+  /*
+  Returns whether or not the string is not empty.
+
+    String.isNotEmpty("a") == true
+    String.isNotEmpty("   ") == false
```
Comment (Sija) [inline core/source/String.mint:45]:
```suggestion
    String.isNotEmpty("   ") == true
```

File: src/formatters/enum_id.cr [hunk 33]
```diff
@@ -2,7 +2,11 @@ module Mint
   class Formatter
     def format(node : Ast::EnumId)
       expressions =
-        unless node.expressions.empty?
+        if node.expressions.empty?
+          ""
+        elsif node.new_line?
+          "(\n#{indent(format(node.expressions, ",\n"))}\n)"
```
Comment (Sija) [inline src/formatters/enum_id.cr:8]:
```suggestion
        case
        when node.expressions.empty?
          ""
        when node.new_line?
          "(\n#{indent(format(node.expressions, ",\n"))}\n)"
```

**Review by Sija — APPROVED** [review]
